renderer/FrameBuffer.cpp

```diff
@@ -464,3 +464,221 @@ void FB_TogglePrimary( bool on ) {
 	else
 		LeavePrimary();
 }
+
+frameBuffers_t frameBuffers;
+
+void FB_InitFrameBuffers() {
+	FB_ShutdownFrameBuffers();
+
+	FrameBuffer *backBuffer = FrameBuffer::BackBuffer();
+	frameBuffers.finalOutput = backBuffer;
+
+	if( r_useFbo.GetBool() ) {
+		
+	} else {
+		frameBuffers.primary = backBuffer;
+		frameBuffers.resolve = backBuffer;
+		frameBuffers.lightgem = backBuffer;
+	}
+}
+
+void FB_ShutdownFrameBuffers() {
+	for( auto it : frameBuffers.allFrameBuffers ) {
+		delete it;
+	}
+	frameBuffers.allFrameBuffers.clear();
+
+	frameBuffers.primary = nullptr;
+	frameBuffers.resolve = nullptr;
+	frameBuffers.lightgem = nullptr;
+	frameBuffers.backBuffer = nullptr;
+	frameBuffers.finalOutput = nullptr;
+	frameBuffers.currentDraw = nullptr;
+	frameBuffers.currentRead = nullptr;
+}
+
+FrameBuffer::FrameBuffer( GLuint fbo, GLuint width, GLuint height, int msaa )
+	: fbo( fbo ), colorBuffer( 0 ), depthStencilBuffer( 0 ), width( width ), height( height ), msaa( msaa ) {
+}
+
+
+FrameBuffer::~FrameBuffer() {
+	if( fbo != 0 ) {
+		qglDeleteFramebuffers( 1, &fbo );
+	}
+	if( colorBuffer != 0 ) {
+		qglDeleteRenderbuffers( 1, &colorBuffer );
+	}
+	if( depthStencilBuffer != 0 ) {
+		qglDeleteRenderbuffers( 1, &depthStencilBuffer );
+	}
+}
+
+FrameBuffer * FrameBuffer::Create( GLuint width, GLuint height, int msaa ) {
+	GLuint fbo = 0;
+	qglGenFramebuffers( 1, &fbo );
+	if( fbo == 0 ) {
+		common->FatalError( "Failed to allocate framebuffer object" );
+	}
+
+	FrameBuffer *frameBuffer = new FrameBuffer( fbo, width, height, msaa );
+	frameBuffers.allFrameBuffers.push_back( frameBuffer );
+	common->Printf( "FBO created: %dx%d, AA = %d\n", width, height, msaa );
+	return frameBuffer;
+}
+
+FrameBuffer * FrameBuffer::BackBuffer() {
+	if( frameBuffers.backBuffer == 0 ) {
+		frameBuffers.backBuffer = new FrameBuffer( 0, glConfig.vidWidth, glConfig.vidHeight, 0 );
+		frameBuffers.allFrameBuffers.push_back( frameBuffers.backBuffer );
+		common->Printf( "Backbuffer dimensions: %dx%d\n", frameBuffers.backBuffer->width, frameBuffers.backBuffer->height );
+	}
+	return frameBuffers.backBuffer;
+}
+
+void FrameBuffer::CreateColorBuffer() {
+	if( colorBuffer == 0 ) {
+		qglGenRenderbuffers( 1, &colorBuffer );
+		if( colorBuffer == 0 ) {
+			common->FatalError( "Failed to allocate color render buffer" );
+		}
+	}
+	qglBindRenderbuffer( GL_RENDERBUFFER, colorBuffer );
+	if( msaa > 1 ) {
+		qglRenderbufferStorageMultisample( GL_RENDERBUFFER, msaa, GL_RGBA, width, height );
```

Comment:
Since we are eager to support EXT_framebuffer extensions for 2.07, it is probably worth checking if this function pointer is not NULL, with fallback to creating non-multisampled renderbuffer if function is missing.
This is one tiny fix to line 547 =)

EDIT: or maybe it would be better to do this check in `FrameBuffer::Create` and change msaa right there.

Reply:
Yeah, the fallbacks are still TODO; there's also an ARB extension covering framebuffers including MSAA. My hope is that the function signatures are identical and we can implement most of the fallbacks during loading of the function pointers, but will have to see.